File: CharChem/Python/CharChem/math/Point.py
class Point:
	""" 2D Point (or vector) object """
	__slots__ = ('x', 'y')
	def __init__(self, x = 0.0, y = 0.0):
		self.x = x
		self.y = y
	def init(self, newX, newY):
		""" Reusing a point instance with new values.
		 Returns this point object with new x, y values. """
		self.x = newX
		self.y = newY
		return self
	def fromPoint(self, pt):
		""" Copying a point from another object """
		self.x = pt.x
		self.y = pt.y
	def clone(self):
		""" Point cloning """
		return Point(self.x, self.y)
	@staticmethod
	def is0(a):
		return abs(a) < 0.001
	def __eq__(self, pt):
		return is0(self.x - pt.x) and is0(self.y - pt.y)
	def iaddn(self, x, y):
		""" Point operator += (x, y) """
		self.x += x
		self.y += y
		return self
	def __iadd__(self, pt):
		self.x += pt.x
		self.y += pt.y
		return self
	def addn(self, x, y):
		""" Add external numbers. Point operator + (x, y) """
		return Point(self.x + x, self.y + y)
	def __add__(self, pt):
		""" Add external point. Point operator + (Point) """
		return Point(self.x + pt.x, self.y + pt.y)
	def isubn(self, x, y):
		""" Subtraction internal numbers. Point operator -= (x, y) """
		self.x -= x
		self.y -= y
		return self
	def __isub__(self, pt):
		""" subtraction internal (Point) """
		self.x -= pt.x
		self.y -= pt.y
		return self
	def subn(self, x, y):
		""" Point operator - (x, y) """
		return Point(self.x - x, self.y - y)
	def __sub__(self, pt):
		""" Point operator - (Point) """
		return Point(self.x - pt.x, self.y - pt.y)
	def __neg__(self):
		return Point(-self.x, -self.y)
	def iminn(self, x1, y1):
		""" min internal numbers """
		self.x = min(self.x, x1)
		self.y = min(self.y, y1)
		return self
	def imin(self, pt):
		""" min internal (Point) """
		return iminn(pt.x, pt.y)
	def imaxn(self, x1, y1):
		""" max internal numbers """
		self.x = max(self.x, x1)
		self.y = max(self.y, y1)
		return self
	def imax(self, pt):
		""" max internal (Point) """
		return imaxn(pt.x, pt.y)
	def ineg(self):
		""" negative internal: pt = -pt """
		self.x = -self.x
		self.y = -self.y
		return self
	def __imul__(self, k):
		""" internal multiply by coefficient """
		self.x *= k
		self.y *= k
		return self
	def __mul__(self, k):
		return Point(self.x * k, self.y * k)
	def __rmul__(self, k):
		return Point(k * self.x, k * self.y)
	def lengthSqr(self):
		""" square of length """
		return self.x ** 2 + self.y ** 2
	def length(self):
		""" Length """
		return sqrt(lengthSqr())
	def distSqrn(self, x1, y1):
		""" Square of distance to point, defined by numbers """
		return (self.x - x1) ** 2 + (self.y - y1) ** 2
	def distSqr(self, pt):
		""" Square of distance to point """
		return self.distSqrn(pt.x, pt.y)
	def dist(self, pt):
		""" Distance to point """
		return sqrt(self.distSqr(pt))

File: CharChem/TypeScript/CharChem/math/Point.ts
// 2D Point (or vector) object
export class Point {
	public x: number;
	public y: number;
	public constructor(x: number = 0.0, y: number = 0.0) {
		this.x = x;
		this.y = y;
	}
	public init(newX: number, newY: number): Point {
		this.x = newX;
		this.y = newY;
		return this;
	}
	public fromPoint(pt: Point) {
		this.x = pt.x;
		this.y = pt.y;
	}
	public clone(): Point {
		return new Point(this.x, this.y);
	}
	public static is0(a: number): boolean {
		return Math.abs(a) < 0.001;
	}
	public eq(pt: Point): boolean {
		return Point.is0(this.x - pt.x) && Point.is0(this.y - pt.y);
	}
	public iaddn(x: number, y: number): Point {
		this.x += x;
		this.y += y;
		return this;
	}
	public iadd(pt: Point): Point {
		this.x += pt.x;
		this.y += pt.y;
		return this;
	}
	public addn(x: number, y: number): Point {
		return new Point(this.x + x, this.y + y);
	}
	// Add external point. Point operator + (Point)
	public add(pt: Point): Point {
		return new Point(this.x + pt.x, this.y + pt.y);
	}
	public isubn(x: number, y: number): Point {
		this.x -= x;
		this.y -= y;
		return this;
	}
	// subtraction internal (Point)
	public isub(pt: Point): Point {
		this.x -= pt.x;
		this.y -= pt.y;
		return this;
	}
	public subn(x: number, y: number): Point {
		return new Point(this.x - x, this.y - y);
	}
	// Point operator - (Point)
	public sub(pt: Point): Point {
		return new Point(this.x - pt.x, this.y - pt.y);
	}
	public neg(): Point {
		return new Point(-this.x, -this.y);
	}
	public iminn(x1: number, y1: number): Point {
		this.x = Math.min(this.x, x1);
		this.y = Math.min(this.y, y1);
		return this;
	}
	public imin(pt: Point): Point {
		return this.iminn(pt.x, pt.y);
	}
	public imaxn(x1: number, y1: number): Point {
		this.x = Math.max(this.x, x1);
		this.y = Math.max(this.y, y1);
		return this;
	}
	public imax(pt: Point): Point {
		return this.imaxn(pt.x, pt.y);
	}
	public ineg(): Point {
		this.x = -this.x;
		this.y = -this.y;
		return this;
	}
	// internal multiply by coefficient
	public imul(k: number): Point {
		this.x *= k;
		this.y *= k;
		return this;
	}
	public mul(k: number): Point {
		return new Point(this.x * k, this.y * k);
	}
	public rmul(k: number): Point {
		return new Point(k * this.x, k * this.y);
	}
	public lengthSqr(): number {
		return this.x ** 2 + this.y ** 2;
	}
	public length(): number {
		return Math.sqrt(this.lengthSqr());
	}
	public distSqrn(x1: number, y1: number): number {
		return (this.x - x1) ** 2 + (this.y - y1) ** 2;
	}
	public distSqr(pt: Point): number {
		return this.distSqrn(pt.x, pt.y);
	}
	public dist(pt: Point): number {
		return Math.sqrt(this.distSqr(pt));
	}
	public fromRad(radAngle: number): Point {
		this.x = Math.cos(radAngle);
		this.y = Math.sin(radAngle);
		return this;
	}
	public fromDeg(degAngle: number): Point {
		return this.fromRad(degAngle * 0.017453292519943295);
	}
	public itranspose(): Point {
		let tmp: number = this.x;
		this.x = this.y;
		this.y = tmp;
		return this;
	}
	public transpose(): Point {
		return new Point(this.y, this.x);
	}
	public static toa(value: number): string {
		return String(Math.round(value * 1000) / 1000);
	}
	public toString(): string {
		return '(' + Point.toa(this.x) + ', ' + Point.toa(this.y) + ')';
	}
	public polarAngle(): number {
		if (this.x === 0) {
			if (this.y === 0) {
				return 0;
			}
			return this.y > 0 ? Math.PI / 2 : -Math.PI / 2;
		}
		return Math.atan2(this.y, this.x);
	}
}

